contents/kotlin/NullSafety.md

```diff
@@ -0,0 +1,121 @@
+```java
+String a = null;
+if(a != null) {
+    if(a.length > 5) {
+        //do something
+    }
+}
+```
+And of course, that's all fine, until we want to do something more complex. Say Bob belongs to a department, and we want to get the name of the department manager. That will look like this:
+```java
+String managerName = bob.department.manager.name;
+```
+
+Because each variable can be `null`, to prevent the NPE we put the code in the following code block:
+```java
+if(bob != null) {
+    Department department = bob.department;
+    if(department != null) {
+        Employee manager = department.manager;
+        if(manager != null) {
+            String name = manager.name;
+            if(name != null) {
+                //do something
+            }
+        }
+    }
+}
+```
+The deep-nested `if` statement adds to the verbosity of our code. Let's see how Kotlin deals with this issue.
+
+# Nullable and Non-nullable type
+In Kotlin, a type can be nullable or non-nullable, determined by the presence of a `?`. For example, an object of type `String` is non-nullable, while an object of type `String?` is nullable.  
+
+As the compiler catches `null` assignments to non-nullable object, the following would result in compilation error.
+```Kotlin
+var firstString: String = "hello world"
+firstString = null //compilation error
+```
+In comparison, the following assignment to a nullable type is allowed
+```Kotlin
+var secondString: String? = "hello world"
+secondString = null //okay
+```
+
+In the first case, we can safely call `firstString.length` without having to worry about a NPE because `firstString` can never be `null`. In the second case, `secondString` can potentially be `null`, so `secondString.length` will result in a compilation error as the compiler see the danger of such statement and blocks it early.
+
+# Operators in Kotlin
+Although non-nullable type is a strong feature in Kotlin, the interoperability with Java means that we have to use variables as nullable type sometimes. In the previous section, we seem to have hit an obstacle as the compiler blocks the call to `secondString.length`. In this section we look at some ways of overcoming this problem.
+## Safe call operator
+Represented by `?.`, the safe call operator is used in this way  
+```kotlin
+secondString?.length
+```
+This returns the length of `secondString` if `secondString` is not `null`, and `null` otherwise.
+
+Now we can chain like this
+```Kotlin
+bob?.department?.manager?.name
+```
+This chain will return `null` if any of the variables inside the chain is `null`.
+
+## Elvis Operators
+Represented by `?:`, the Elvis operator is used in this way
+```kotlin
+val length = secondString?.length ?: -1
+```
+If the expression to the left of `?:` is not null, the Elvis operator will return it as it is; else it will return a default value supplied (-1 in this case).
+
+We also notice the use of safe call operator together with Elvis operator in the same statement.
```

Comment:
Might want to add a "(`?.`)" next to "safe call operator" since readers may be new to the concept and forget that it was just introduced in the previous subsection.